kotlinx-coroutines-core/common/src/flow/operators/Merge.kt

```diff
@@ -11,64 +11,54 @@ package kotlinx.coroutines.flow
  * Most likely, suspending operation in [map] operator will be sufficient and linear transformations are much easier to reason about.
  */
 @FlowPreview
-public fun <T, R> Flow<T>.flatMapConcat(transform: suspend (value: T) -> Flow<R>): Flow<R> = flow {
-    collect { value ->
-        transform(value).collect { innerValue ->
-            emit(innerValue)
-        }
-    }
-}
+public fun <T, R> Flow<T>.flatMapConcat(transform: suspend (value: T) -> Flow<R>): Flow<R> =
+    map(transform).flattenConcat()
 
 /**
- * Transforms elements emitted by the original flow by applying [transform], that returns another flow, and then merging and flattening these flows.
+ * Transforms elements emitted by the original flow by applying [transform], that returns another flow,
+ * and then merging and flattening these flows.
+ *
+ * This operator calls [transform] *sequentially* and then merges the resulting flows with a [concurrency]
+ * limit on the number of concurrently collected flows.
+ * It is a shortcut for `map(transform).flattenMerge(concurrency)`.
+ * See [flattenMerge] for details.
  *
  * Note that even though this operator looks very familiar, we discourage its usage in a regular application-specific flows.
  * Most likely, suspending operation in [map] operator will be sufficient and linear transformations are much easier to reason about.
  *
- * [bufferSize] parameter controls the size of backpressure aka the amount of queued in-flight elements.
- * [concurrency] parameter controls the size of in-flight flows, at most [concurrency] flows are collected at the same time.
+ * @param concurrency controls the number of in-flight flows, at most [concurrency] flows are collected
+ * at the same time. By default it is equal to [DEFAULT_CONCURRENCY].
  */
 @FlowPreview
-public fun <T, R> Flow<T>.flatMapMerge(concurrency: Int = 16, bufferSize: Int = 16, transform: suspend (value: T) -> Flow<R>): Flow<R> {
-    require(bufferSize >= 0) { "Expected non-negative buffer size, but had $bufferSize" }
-    require(concurrency >= 0) { "Expected non-negative concurrency level, but had $concurrency" }
-    return flow {
-        coroutineScope {
-            val semaphore = Channel<Unit>(concurrency)
-            val flatMap = SerializingFlatMapCollector(this@flow, bufferSize)
-            collect { outerValue ->
-                // TODO real semaphore (#94)
-                semaphore.send(Unit) // Acquire concurrency permit
-                val inner = transform(outerValue)
-                launch {
-                    try {
-                        inner.collect { value ->
-                            flatMap.emit(value)
-                        }
-                    } finally {
-                        semaphore.receive() // Release concurrency permit
-                    }
-                }
-            }
-        }
-    }
-}
+public fun <T, R> Flow<T>.flatMapMerge(
```

Comment:
This operator is a bit "inconsistent" in both implementation and documentation (if e.g. compared to `flowOn`):
```
// Not fused, two buffers
.buffer(1).flatMapMerge { flowOf(it) }.buffer(2)

// Fused, one buffer
.buffer(1).flowOn(Dispatchers.Default).buffer(2)

// Fused partially
.buffer(1).flatMapMerge { flowOf(it) }.buffer(2).flowOn(Dispatchers.Default).buffer(3)
```

I am fine with inconsistency, but the fact that flatMap is fused with consecutive buffered ops should be documented

Reply:
Added sections in docs on fusion:

### Operator fusion

Applications of [flowOn], [buffer], [produceIn], and [broadcastIn] _after_ this operator are fused with its concurrent merging so that only one properly configured channel is used for execution of merging logic.